### PR TITLE
xwayland: add server and basic window functionality

### PR DESCRIPTION
**Pull request by alex-courtis**
## dwl: Great News

So happy to find this project! I'm a massive dwm fan, who knows it is soon time to move to wayland.

Using the mature wlroots is the right thing to do, even if it's not quite suckless. I even started such a project, but never followed through: https://github.com/alex-courtis/wdwm

I am hopeful that the preserved structure of dwl will result in many dwm patches still applying, or applying with little change.

Happy to help in any way I can, so I picked a task off the feature list in the readme.

cheers,
Alex

## XWayland

These changes add the X server and dwl functionality for basic X clients. It's been tested with st, chromium and intellij, at scale 1 and 2.

I don't know your preferences: merge incrementally or keep a long running branch until feature complete. This PR is here, you can decide.

Next:
- Dropdown menu support
- X selection to wayland selection
- Focus passing between surfaces from the same process (for Jetbrains)
- X properties providing
- Floating windows

Remaining functionality will be added at the same time dwl implements e.g. fullscreen

### CODE REVIEW

File: dwl.c [hunk 1]
```diff
@@ -64,11 +66,15 @@ typedef struct {
...
-	struct wlr_xdg_surface *xdg_surface;
+	union {
+		struct wlr_xdg_surface *xdg_surface;
+		struct wlr_xwayland_surface *xwayland_surface;
+	};
 	struct wl_listener map;
 	struct wl_listener unmap;
 	struct wl_listener destroy;
 	struct wlr_box geom;  /* layout-relative, includes border */
+	int isxdg;
```
Comment (alex-courtis) [inline dwl.c:77]: This could become an enum, when we add layer shell.

File: dwl.c [hunk 2]
```diff
@@ -271,10 +281,15 @@ applyrules(Client *c)
...
-	if (!(title = c->xdg_surface->toplevel->title))
-		title = broken;
+	if (c->isxdg) {
+		if (!(appid = c->xdg_surface->toplevel->app_id))
+			appid = broken;
+		if (!(title = c->xdg_surface->toplevel->title))
+			title = broken;
+	} else {
+		if (!(title = c->xwayland_surface->title))
+			title = broken;
```
Comment (alex-courtis) [inline dwl.c:291]: xwayland does not have an app_id

File: dwl.c [hunk 3]
```diff
@@ -328,9 +343,14 @@ buttonpress(struct wl_listener *listener, void *data)
...
-					cursor->x - c->geom.x - c->bw,
-					cursor->y - c->geom.y - c->bw, NULL, NULL);
+			if (c->isxdg)
+				surface = wlr_xdg_surface_surface_at(c->xdg_surface,
+						cursor->x - c->geom.x - c->bw,
+						cursor->y - c->geom.y - c->bw, NULL, NULL);
+			else
+				surface = wlr_surface_surface_at(c->xwayland_surface->surface,
+						cursor->x - c->geom.x - c->bw,
+						cursor->y - c->geom.y - c->bw, NULL, NULL);
```
Comment (alex-courtis) [inline dwl.c:353]: Cases like these intentionally invoke different methods, as the xdg version has more functionality.

File: dwl.c [hunk 4]
```diff
@@ -1550,6 +1643,7 @@ main(int argc, char *argv[])
 	run(startup_cmd);
 
 	/* Once wl_display_run returns, we shut down the server. */
+	wlr_xwayland_destroy(xwayland);
```
Comment (alex-courtis) [inline dwl.c:1646]: This is null safe.

File: dwl.c [hunk 5]
```diff
@@ -64,11 +66,15 @@ typedef struct {
 	struct wl_list link;
 	struct wl_list flink;
 	struct wl_list slink;
-	struct wlr_xdg_surface *xdg_surface;
+	union {
+		struct wlr_xdg_surface *xdg_surface;
+		struct wlr_xwayland_surface *xwayland_surface;
+	};
```
Comment (alex-courtis) [inline dwl.c:72]: The anonymous-union-and-a-boolean seemed the simplest approach, and sway is using it.

Happy to change if anything better is suggested.

File: dwl.c [hunk 6]
```diff
@@ -892,8 +957,9 @@ pointerfocus(Client *c, struct wlr_surface *surface, double sx, double sy,
 		uint32_t time)
 {
 	/* Use top level surface if nothing more specific given */
-	if (c && !surface)
-		surface = c->xdg_surface->surface;
+	if (c && !surface) {
+		surface = WLR_SURFACE(c);
+	}
```
Comment (theotherjimmy) [inline dwl.c:962]: It looks like this repo prefers to omit braces when an if contains a single statement. This change does not match that style.

Comment (alex-courtis) [reply dwl.c:961]: Fixed; thanks for spotting it.

**Comment by theotherjimmy** [comment]
Thanks for getting back to the review! It looks like your commits are all the same message. I'm not sure what the policy is for this repo (note that they may prefer you not rebase/force push more than they care about commit messages), but I encourage commit messages that describe what's going on in the commit. Something like "Remove braces from single-statement if" is what I would write for that last commit.

File: dwl.c (hunk 2 above)
Comment (djpohly) [reply dwl.c:291]: It does appear to still have instance/class though

**Comment by djpohly** [comment]
Thanks @alex-courtis for the work!  Sorry I couldn't get to major dwl stuff until now (teaching + research + pandemic = crazy), but this looks like a great starting point for XWayland.

**Comment by alex-courtis** [comment]
> Thanks @alex-courtis for the work! Sorry I couldn't get to major dwl stuff until now (teaching + research + pandemic = crazy), but this looks like a great starting point for XWayland.

No worries. More xwayland PRs incoming.

File: dwl.c (hunk 2 above)
Comment (alex-courtis) [reply dwl.c:291]: > It does appear to still have instance/class though

We can add those when we use them e.g. rules based on class.

**Pull request merged**